### PR TITLE
add typescript definitions for OffsetDateTime and OffsetTime

### LINKED ISSUE #440: OffsetDateTime and OffsetTime missing in typescript definitions
OffsetDateTime and OffsetTime, recently added in @js-joda/core@3.0.0 are still missing in the typescript definition files.

Thanks to @xenoterracide, reported this at #432

### PR DESCRIPTION
**Pull request by jonfreedman**
Fixes #440

### CODE REVIEW

**Comment by pithu** [comment]
Hi @jonfreedman, 

great, that you taking the initiative! 

Could you add some test to `packages/core/test/typescript_definitions/`, similar to the existing once ? As we decided to accept typescript PR's only with a basic test coverage. Would be great.

**Comment by jonfreedman** [comment]
@pithu thanks for the hit, that flagged quite a few issues after just transposing from javadoc, can you take a look now please?

**Comment by InExtremaRes** [comment]
Hi. Thank you for the types declarations. I'm not a member of the maintenance team but I've collaborated to this types a few times before, so I hope there are no troubles that I looked into this and giving some opinion.

There are a few details I notice in a quick view:

- Both classes lack the types for the static properties `MAX` and `MIN`.
- Both classes should `extends Temporal` and `implements TemporalAdjuster`, mainly for consistency.
- `OffsetTime.prototype.adjustInto` should accept and return `Temporal` not `TemporalAccesor`.

File: packages/core/dist/js-joda.d.ts [hunk 1]
```diff
@@ -1549,6 +1550,183 @@ export class YearMonth extends Temporal implements TemporalAdjuster {
...
+    range(field: TemporalField): ValueRange
+    toEpochSecond(): number
+    toJSON(): string;
+    toInstant(): Instant
+    toLocalDate(): LocalDate
+    toLocalDateTime(): LocalDateTime
+    toLocalTime(): LocalTime
+    toOffsetTime(): OffsetTime
+    toString(): string
+    toOffsetDateTime(): OffsetDateTime
```
Comment (pithu) [inline packages/core/dist/js-joda.d.ts:1642]: There is no method `toOffsetDateTime` or am i wrong ?

File: packages/core/dist/js-joda.d.ts [hunk 2]
```diff
@@ -1549,6 +1550,183 @@ export class YearMonth extends Temporal implements TemporalAdjuster {
...
+    truncatedTo(unit: TemporalUnit): OffsetDateTime
```
Comment (pithu) [inline packages/core/dist/js-joda.d.ts:1643]: dito, `truncatedTo` is not in the API

File: packages/core/dist/js-joda.d.ts [hunk 3]
```diff
@@ -1549,6 +1550,183 @@ export class YearMonth extends Temporal implements TemporalAdjuster {
...
+    until(endExclusive: Temporal, unit: TemporalUnit): number
+    with(adjuster: TemporalAdjuster): OffsetDateTime
+    with(field: TemporalField, newValue: number): OffsetDateTime
```
Comment (pithu) [inline packages/core/dist/js-joda.d.ts:1646]: Unfortunately implementation of `with` is still missing. must be implemented.

File: packages/core/dist/js-joda.d.ts [hunk 4]
```diff
@@ -1549,6 +1550,183 @@ export class YearMonth extends Temporal implements TemporalAdjuster {
...
+    query<T>(query: TemporalQuery<T>): T
+    range(field: TemporalField): ValueRange
+    toEpochSecond(date: LocalDate): number
+    toJSON(): string;
+    toLocalTime(): LocalTime
+    toString(): string
+    truncatedTo(unit: TemporalUnit): OffsetTime
+    until(endExclusive: Temporal, unit: TemporalUnit): number
+    with(adjuster: TemporalAdjuster): OffsetTime
+    with(field: TemporalField, newValue: number): OffsetTime
```
Comment (pithu) [inline packages/core/dist/js-joda.d.ts:1721]: dito `with` is still missing in `OffsetTime`

**Review by pithu — CHANGES_REQUESTED** [review]
Cool stuff, but i found some mismatches, i can't merge it like that. Could you fix that issues ?

Beside that, can you have a look at the comments of @InExtremaRes. Sounds reasonable to me, he contributed a lot to the js-joda typescript definitions.

File: packages/core/dist/js-joda.d.ts (hunk 1 above)
Comment (jonfreedman) [reply packages/core/dist/js-joda.d.ts:1642]: Yeah this should be on ZonedDateTime, not sure how that happened...

File: packages/core/dist/js-joda.d.ts (hunk 2 above)
Comment (jonfreedman) [reply packages/core/dist/js-joda.d.ts:1643]: It seems to be in both `OffsetTime.js` and `OffsetDateTime.js`

* https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/OffsetDateTime.html#truncatedTo(java.time.temporal.TemporalUnit)
* https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/OffsetTime.html#truncatedTo(java.time.temporal.TemporalUnit)

File: packages/core/dist/js-joda.d.ts (hunk 4 above)
Comment (jonfreedman) [reply packages/core/dist/js-joda.d.ts:1721]: removed

File: packages/core/dist/js-joda.d.ts (hunk 3 above)
Comment (jonfreedman) [reply packages/core/dist/js-joda.d.ts:1646]: removed

**Comment by jonfreedman** [comment]
It looks like `with` must be left in the typescript type in order to implement Temporal - you can see it's defined on various other types e.g. `Instant` and `LocalDate`

File: packages/core/dist/js-joda.d.ts (hunk 2 above)
Comment (pithu) [reply packages/core/dist/js-joda.d.ts:1644]: my bad

File: packages/core/dist/js-joda.d.ts (hunk 3 above)
Comment (InExtremaRes) [reply packages/core/dist/js-joda.d.ts:1647]: @pithu `with` should be inherited from class `Temporal`:

https://github.com/js-joda/js-joda/blob/166f123d92ccd9d4ebf1ef86fe992e6c233791aa/packages/core/src/temporal/Temporal.js#L291-L297

Isn't it? `OffsetDateTime` and `OffsetTime` both extends that class.

Comment (jonfreedman) [reply packages/core/dist/js-joda.d.ts:1647]: Yes and it's exercised in the tests

**Review by pithu — APPROVED** [review]
LGTM

**Pull request merged**